sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/UnsafeArrayDataBenchmark.scala

```diff
@@ -0,0 +1,224 @@
+        n += 1
+      }
+    }
+
+    val benchmark = new Benchmark("Write UnsafeArrayData", count * iters)
+    benchmark.addCase("Int")(writeIntArray)
+    benchmark.addCase("Double")(writeDoubleArray)
+    benchmark.run
+    /*
+    Java HotSpot(TM) 64-Bit Server VM 1.8.0_92-b14 on Mac OS X 10.10.4
+    Intel(R) Core(TM) i5-5257U CPU @ 2.70GHz
+
+    Write UnsafeArrayData:                   Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
+    ------------------------------------------------------------------------------------------------
+    Int                                            135 /  180        154.9           6.5       1.0X
+    Double                                         195 /  300        107.8           9.3       0.7X
+    */
+  }
+
+  def getPrimitiveArray(iters: Int): Unit = {
+    val count = 1024 * 1024 * 12
+    val rand = new Random(42)
+
+    var intTotalLength: Int = 0
+    val intPrimitiveArray = Array.fill[Int](count) { rand.nextInt }
+    val intEncoder = ExpressionEncoder[Array[Int]].resolveAndBind()
+    val intUnsafeArray = intEncoder.toRow(intPrimitiveArray).getArray(0)
+    val readIntArray = { i: Int =>
+      var n = 0
+      while (n < iters) {
+        intTotalLength += intUnsafeArray.toIntArray.length
+        n += 1
+      }
+    }
+
+    var doubleTotalLength: Int = 0
+    val doublePrimitiveArray = Array.fill[Double](count) { rand.nextDouble }
+    val doubleEncoder = ExpressionEncoder[Array[Double]].resolveAndBind()
+    val doubleUnsafeArray = doubleEncoder.toRow(doublePrimitiveArray).getArray(0)
+    val readDoubleArray = { i: Int =>
+      var n = 0
+      while (n < iters) {
+        doubleTotalLength += doubleUnsafeArray.toDoubleArray.length
+        n += 1
+      }
+    }
+
+    val benchmark = new Benchmark("Get primitive array from UnsafeArrayData", count * iters)
+    benchmark.addCase("Int")(readIntArray)
+    benchmark.addCase("Double")(readDoubleArray)
+    benchmark.run
+    /*
+    Java HotSpot(TM) 64-Bit Server VM 1.8.0_92-b14 on Mac OS X 10.10.4
+    Intel(R) Core(TM) i5-5257U CPU @ 2.70GHz
+
+    Get primitive array from UnsafeArrayData: Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)  Relative
+    ------------------------------------------------------------------------------------------------
+    Int                                             80 /  151        783.4           1.3       1.0X
+    Double                                         208 /  366        302.8           3.3       0.4X
+    */
+  }
+
+  def putPrimitiveArray(iters: Int): Unit = {
+    val count = 1024 * 1024 * 12
+    val rand = new Random(42)
+
+    val intPrimitiveArray = Array.fill[Int](count) { rand.nextInt }
+    var intUnsafeArray: UnsafeArrayData = null
+    val createIntArray = { i: Int =>
+      var n = 0
+      while (n < iters) {
+        intUnsafeArray = UnsafeArrayData.fromPrimitiveArray(intPrimitiveArray)
```

Comment:
this may get optimized , let's do
`totalLength += UnsafeArrayData.fromPrimitiveArray(intPrimitiveArray).numElements`